plugins/module_utils/data_set.py

```diff
@@ -1067,25 +1111,53 @@ def _catalog_vsam(name, volumes):
         data_set_name = name.upper()
         success = False
         command_rc = 0
-        for data_set_type in ["", "LINEAR", "INDEXED", "NONINDEXED", "NUMBERED"]:
-            if data_set_type != "INDEXED":
-                command = DataSet._VSAM_CATALOG_COMMAND_NOT_INDEXED.format(
-                    data_set_name,
-                    DataSet._build_volume_string_idcams(volumes),
-                    data_set_type,
-                )
-            else:
-                command = DataSet._VSAM_CATALOG_COMMAND_INDEXED.format(
-                    data_set_name,
-                    DataSet._build_volume_string_idcams(volumes),
-                    data_set_type,
-                )
-            command_rc, stdout, stderr = module.run_command(
-                "mvscmdauth --pgm=idcams --sysprint=* --sysin=stdin", data=command
+        command = ""
+
+        # In order to catalog a uncataloged data set, we can't rely on LISTCAT
+        # so using the VTOC entries we can make some assumptions of if the data set
+        # is indexed, linear etc.
+        ds_vtoc_data_entry = vtoc.get_data_set_entry(name + ".DATA", volumes[0])
+        ds_vtoc_index_entry = vtoc.get_data_set_entry(name + ".INDEX", volumes[0])
+
+        if ds_vtoc_data_entry and ds_vtoc_index_entry:
+            data_set_type_vsam = "INDEXED"
+        else:
+            data_set_type_vsam = "NONINDEXED"
+
+        if data_set_type_vsam != "INDEXED":
+            command = DataSet._VSAM_CATALOG_COMMAND_NOT_INDEXED.format(
+                data_set_name,
+                DataSet._build_volume_string_idcams(volumes),
+                data_set_type_vsam,
+            )
+        else:
+            command = DataSet._VSAM_CATALOG_COMMAND_INDEXED.format(
+                data_set_name,
+                DataSet._build_volume_string_idcams(volumes),
+                data_set_type_vsam,
             )
+
+        command_rc, stdout, stderr = module.run_command(
```

Comment:
This is not a deal breaker but why to use module.run_command to execute idcams if we have module_utils/mvs_cmd.idcams?

Reply:
Note: Not requesting to change this

Reply:
You are correct it needs to be standardized into one way of interacting with IDCAMs. I left this untouched because we have a item in plan next quarter to find all these and standardize on one way issue #437 , I would then even go as far as looking at mvs_cmd's usage of run_command because its been a bit[ problematic in terms of heap](https://github.com/ansible/ansible/issues/80379) , we might explore the ZOAU Python APIs instead as well. Good point.

Reply:
Thanks for the info